File: src/main/java/caelum/focalworks/casting/actions/OpRigWrite.java
package caelum.focalworks.casting.actions;

import at.petrak.hexcasting.api.casting.castables.ConstMediaAction;
import at.petrak.hexcasting.api.casting.eval.CastingEnvironment;
import at.petrak.hexcasting.api.casting.eval.OperationResult;
import at.petrak.hexcasting.api.casting.eval.vm.CastingImage;
import at.petrak.hexcasting.api.casting.eval.vm.SpellContinuation;
import at.petrak.hexcasting.api.casting.iota.Iota;
import static at.petrak.hexcasting.api.casting.OperatorUtils.getItemEntity;
import static at.petrak.hexcasting.api.casting.OperatorUtils.getList;

import at.petrak.hexcasting.api.casting.iota.IotaType;
import at.petrak.hexcasting.api.casting.iota.ListIota;
import at.petrak.hexcasting.api.casting.mishaps.Mishap;
import at.petrak.hexcasting.api.misc.MediaConstants;
import at.petrak.hexcasting.api.utils.NBTHelper;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class OpRigWrite implements ConstMediaAction {
    private static final Integer argc = 2;
    @Override
    public int getArgc() {
        return argc;
    }
    @Override
    public long getMediaCost() {
        return 20 * MediaConstants.DUST_UNIT;
    }

    @Override
    public @NotNull List<Iota> execute(@NotNull List<? extends Iota> args, @NotNull CastingEnvironment env) throws Mishap {
        ItemEntity entity = getItemEntity(args,0,argc);
        Iota hex = new ListIota(getList(args,1,argc));
        ItemStack stack = entity.getItem();
        CompoundTag tag = IotaType.serialize(hex);
        NBTHelper.put(stack,"riggedwrite",tag);
        return List.of();
    }

    @Override
    public @NotNull CostMediaActionResult executeWithOpCount(@NotNull List<? extends Iota> args, @NotNull CastingEnvironment env) throws Mishap {
        return DefaultImpls.executeWithOpCount(this,args,env);
    }

    @Override
    public @NotNull OperationResult operate(@NotNull CastingEnvironment env, @NotNull CastingImage image, @NotNull SpellContinuation continuation) {
        return DefaultImpls.operate(this,env,image,continuation);
    }
}

File: common/src/main/kotlin/io/github/c_caelum/focalworks/casting/actions/spells/OpRigWrite.kt
package io.github.c_caelum.focalworks.casting.actions.spells

import at.petrak.hexcasting.api.casting.castables.ConstMediaAction
import at.petrak.hexcasting.api.casting.eval.CastingEnvironment
import at.petrak.hexcasting.api.casting.getList
import at.petrak.hexcasting.api.casting.iota.Iota
import at.petrak.hexcasting.api.casting.iota.IotaType
import at.petrak.hexcasting.api.casting.iota.ListIota
import at.petrak.hexcasting.api.casting.mishaps.MishapBadOffhandItem
import at.petrak.hexcasting.api.utils.putTag
import at.petrak.hexcasting.xplat.IXplatAbstractions


object OpRigWrite : ConstMediaAction {
    override val argc = 1
    override fun execute(args: List<Iota>, env: CastingEnvironment): List<Iota> {
        val hex = ListIota(args.getList(0,argc))

        val (handStack) = env.getHeldItemToOperateOn {
            val dataHolder = IXplatAbstractions.INSTANCE.findDataHolder(it)
            dataHolder != null && (dataHolder.readIota(env.world) != null || dataHolder.emptyIota() != null)
        }
        // If there are no data holders that are readable, find a data holder that isn't readable
        // so that the error message is more helpful.
            ?: env.getHeldItemToOperateOn {
                val dataHolder = IXplatAbstractions.INSTANCE.findDataHolder(it)
                dataHolder != null
            } ?: throw MishapBadOffhandItem.of(null, "iota.read")

        val datumHolder = IXplatAbstractions.INSTANCE.findDataHolder(handStack)
            ?: throw MishapBadOffhandItem.of(handStack, "iota.read")
        val tag = IotaType.serialize(hex);
        handStack.putTag("riggedread",tag)


        return listOf()
    }
}
